File: server/src/citekit_server/retrieve.py
from __future__ import annotations

import re

from sqlalchemy.orm import Session

from citekit_server.call_log import call_scope
from citekit_server.db import AiModelRow, ChunkRow, KnowledgeBaseRow, SourceRow
from citekit_server.ingest import embedding_model, resolve_auth
from citekit_server.schemas import ChunkOut, SearchHit, SearchIn, SearchOut
from citekit_server.serialize import chunk_to_out
from citekit_server.upstream import embed_texts, rerank_texts
from citekit_server.vectors import search as vector_search


def _keyword_score(query: str, text: str) -> float:
    q = query.strip().lower()
    if not q:
        return 0.0
    t = text.lower()
    score = 0.0
    for token in q.split():
        if token and token in t:
            score += 0.28
    if q in t:
        score += 0.35
    compact = "".join(q.split())
    if re.search(r"[\u4e00-\u9fff]", compact):
        hits = sum(1 for i in range(len(compact) - 1) if compact[i : i + 2] in t)
        if hits:
            score += min(0.5, 0.08 * hits)
    return min(1.0, score)


def search_kb(db: Session, kb: KnowledgeBaseRow, body: SearchIn) -> SearchOut:
    query = (body.query or "").strip()
    if not query:
        return SearchOut(hits=[], message="请输入问题")
    source_q = db.query(SourceRow).filter(SourceRow.kb_id == kb.id, SourceRow.type != "folder")
    sources = source_q.all()
    if body.sourceIds is not None:
        wanted = set(body.sourceIds)
        if not wanted:
            return SearchOut(hits=[], message="请至少勾选一个数据集。")
        sources = [row for row in sources if row.id in wanted]
    source_ids = [row.id for row in sources]
    if not source_ids:
        return SearchOut(hits=[], message="该知识库还没有可检索的数据。请先导入集合并等待就绪。")

    chunks = (
        db.query(ChunkRow)
        .filter(ChunkRow.kb_id == kb.id, ChunkRow.source_id.in_(source_ids))
        .order_by(ChunkRow.position)
        .all()
    )
    if not chunks:
        return SearchOut(hits=[], message="该知识库还没有可检索的数据。请先导入集合并等待就绪。")

    by_id = {row.id: row for row in chunks}
    scored: dict[str, tuple[float, str]] = {}
    mode = body.searchMode or "mix"

    if mode in {"fullText", "mix"}:
        for row in chunks:
            score = _keyword_score(query, f"{row.title} {row.text}")
            if score > 0:
                scored[row.id] = (score, "全文检索")

    if mode in {"embedding", "mix"}:
        try:
            model = embedding_model(db, kb)
            with call_scope(purpose="retrieve", kb_id=kb.id):
                vector = embed_texts(model, resolve_auth(db, model), [query])[0]
            hits = vector_search(kb.id, vector, limit=max(body.limit * 3, 20), source_ids=source_ids)
        except Exception as exc:
            if mode == "embedding":
                return SearchOut(hits=[], message=str(exc) or "语义检索失败")
            hits = []
        for hit in hits:
            payload = hit.payload or {}
            cid = str(payload.get("chunk_id") or hit.id)
            if cid not in by_id:
                continue
            score = float(hit.score or 0)
            prev = scored.get(cid)
            if prev:
                scored[cid] = ((prev[0] + score) / 2, "混合检索")
            else:
                scored[cid] = (score, "语义检索")

    ranked = sorted(scored.items(), key=lambda item: item[1][0], reverse=True)
    ranked = [(cid, score, note) for cid, (score, note) in ranked if score >= body.similarity]
    if not ranked:
        return SearchOut(hits=[], message=f"低于相似度 {body.similarity}，无召回。可调低阈值再试。")

    if body.usingRerank and kb.rerank_model:
        rerank_row = db.get(AiModelRow, kb.rerank_model)
        if rerank_row and rerank_row.type == "rerank":
            pool = ranked[: max(body.limit, 8)]
            docs = [by_id[cid].text for cid, _, _ in pool]
            try:
                with call_scope(purpose="rerank", kb_id=kb.id):
                    scores = rerank_texts(rerank_row, resolve_auth(db, rerank_row), query, docs)
                ranked = [
                    (cid, float(scores[i]), "混合召回后重排")
                    for i, (cid, _, _) in enumerate(pool)
                ]
                ranked.sort(key=lambda item: item[1], reverse=True)
            except Exception:
                pass

    ranked = ranked[: body.limit]
    return SearchOut(
        hits=[
            SearchHit(chunk=chunk_to_out(by_id[cid]), score=round(score, 4), note=note)
            for cid, score, note in ranked
            if cid in by_id
        ]
    )

File: web/src/mock/retrieve.ts
import type { Chunk, RetrievalProfile, SearchConfig } from "../types";

export interface RetrieveInput {
  sliceId?: string;
  sliceIds?: string[];
  sourceIds?: string[];
  query: string;
  profile?: RetrievalProfile;
  search?: SearchConfig;
  warehouse?: string;
}

export interface Hit {
  chunk: Chunk;
  score: number;
  note: string;
}

function scoreText(query: string, text: string): number {
  const q = query.trim().toLowerCase();
  if (!q) return 0;
  const t = text.toLowerCase();
  let s = 0;
  for (const token of q.split(/\s+/)) {
    if (token.length < 1) continue;
    if (t.includes(token)) s += 0.28;
  }
  if (t.includes(q)) s += 0.35;
  return Math.min(1, s);
}

export function retrieve(
  input: RetrieveInput,
  allChunks: Chunk[],
): { hits: Hit[]; message?: string } {
  if (input.sourceIds && input.sourceIds.length === 0) {
    return { hits: [], message: "请至少勾选一个数据集。" };
  }

  let pool = allChunks;
  if (input.sourceIds) {
    pool = pool.filter((c) => c.sourceId && input.sourceIds!.includes(c.sourceId));
  } else if (input.sliceIds?.length) {
    pool = pool.filter((c) => input.sliceIds!.includes(c.sliceId));
  } else if (input.sliceId) {
    pool = pool.filter((c) => c.sliceId === input.sliceId);
  }
  if (pool.length === 0) {
    return { hits: [], message: "该知识库还没有可检索的数据。请先导入集合并等待就绪。" };
  }

  if (input.profile === "filter_first") {
    const wh = input.warehouse?.trim();
    if (!wh) {
      return {
        hits: [],
        message: "empty_after_filter：filter_first 要求提供 warehouse，未静默全表检索。",
      };
    }
    pool = pool.filter((c) => c.warehouse === wh);
    if (pool.length === 0) {
      return { hits: [], message: "empty_after_filter：该仓库下没有匹配行。" };
    }
  }

  const search = input.search;
  const mode = search?.searchMode;
  let ranked: Hit[] = pool
    .map((chunk) => {
      const base = scoreText(input.query, `${chunk.title} ${chunk.text} ${chunk.a ?? ""} ${chunk.sku ?? ""}`);
      const semantic = Math.min(1, base + 0.08);
      const score = mode === "embedding" ? semantic : mode === "fullText" ? base : (base + semantic) / 2;
      const note =
        mode === "embedding"
          ? "语义检索（原型模拟）"
          : mode === "fullText"
            ? "全文检索（原型模拟）"
            : "混合检索（原型模拟）";
      return { chunk, score, note };
    })
    .sort((a, b) => b.score - a.score);

  if (search) {
    ranked = ranked.filter((h) => h.score >= search.similarity);
    if (ranked.length === 0) {
      return { hits: [], message: `低于相似度 ${search.similarity}，无召回。可调低阈值再试。` };
    }
    if (search.usingRerank) {
      ranked = ranked.slice(0, Math.max(search.limit, 8)).map((h, i) => ({
        ...h,
        score: Math.min(1, h.score + 0.12 - i * 0.015),
        note: "混合召回后重排（原型模拟）",
      }));
    }
    ranked = ranked.slice(0, search.limit);
    return { hits: ranked };
  }

  ranked = ranked.filter((h) => h.score > 0);
  if (ranked.length === 0) {
    ranked = pool.map((chunk) => ({
      chunk,
      score: 0.12,
      note: "无关键词命中，返回片内兜底低分块",
    }));
  }

  if (input.profile === "hybrid_rerank_strict") {
    ranked = ranked.slice(0, 8).map((h, i) => ({
      ...h,
      score: Math.min(1, h.score + 0.15 - i * 0.02),
      note: "交叉编码器重排（原型模拟）",
    }));
    ranked = ranked.slice(0, 3);
  } else if (input.profile === "hybrid_rerank_cutoff") {
    ranked = ranked
      .map((h) => ({ ...h, note: "重排后按阈值丢弃（模拟 cutoff=0.45）" }))
      .filter((h) => h.score >= 0.45);
  } else if (input.profile === "keyword_precise") {
    ranked = ranked.filter((h) => h.score >= 0.4).slice(0, 5);
  } else {
    ranked = ranked.slice(0, 5);
  }

  return { hits: ranked };
}
